src/main/scala/org/renci/cam/LookupService.scala

```diff
@@ -0,0 +1,420 @@
+                case lit  => LabeledIRI(res.getResource("subj").getURI, Set(lit.getString))
+              }
+            }
+          )
+          .toSet
+
+        Relation(
+          subjLabeled,
+          predResults.map(_._1).toSet,
+          predResults.map(kv => (kv._1.iri, kv._2)).toMap,
+          objLabeled,
+          Seq[Relation](), // TODO: recurse!
+          results.map(_.getResource("g").getURI).toSet
+        )
+      }
+    } yield relations
+
+  /* Node Norm structures and functions */
+
+  /* Node Norm result structures */
+
+  /** Node Norm identifier */
+  case class NodeNormIdentifier(
+    identifier: String,
+    label: Option[String]
+  ) {
+    val asLabeledIRI = LabeledIRI(identifier, label.toSet)
+  }
+
+  /** Node Norm response */
+  case class NodeNormResponse(
+    id: Option[NodeNormIdentifier],
+    equivalent_identifiers: List[NodeNormIdentifier],
+    `type`: List[BiolinkClass],
+    information_content: Option[Float]
+  )
+
+  /** Retrieve a list of all equivalent identifiers for a queryId.
+    *
+    * @param queryId
+    *   The identifier being queried.
+    * @param nodeNormURL
+    *   The NodeNorm URL (ending with `/get_normalized_nodes`) to query.
+    * @param conflate
+    *   What to set in the `conflate` param to NodeNorm. Currently, may be `true` or `false`.
+    */
+  def getQualifiedIdsFromNodeNorm(queryId: String,
+                                  nodeNormURL: String,
+                                  conflate: String): ZIO[Blocking with Has[BiolinkData], Exception, List[LabeledIRI]] =
+    for {
+      biolinkData <- biolinkData
+      nnUri <- ZIO.fromEither(Uri.fromString(nodeNormURL))
+      uri = nnUri.+?("curie", queryId).+?("conflate", conflate).toString()
+      strResult <- effectBlockingIO(Source.fromURL(uri)).bracketAuto { src =>
+        val lines = src.getLines().mkString
+
+        logger.info(s"Queried ${uri}, got result: ${lines}.")
+
+        ZIO.succeed(lines)
+      }
+      jsonResult <- ZIO.fromEither(parser.parse(strResult))
+      qualifiedIds = (jsonResult \\ queryId) flatMap { res =>
+        // Decode IRIs from NodeNorm into Biolink entities and IRIs.
+        implicit val iriDecoder: Decoder[IRI] = Implicits.iriDecoder(biolinkData.prefixes)
+        implicit val iriEncoder: Encoder[IRI] = Implicits.iriEncoder(biolinkData.prefixes)
+
+        implicit val iriKeyEncoder: KeyEncoder[IRI] = Implicits.iriKeyEncoder(biolinkData.prefixes)
+        implicit val iriKeyDecoder: KeyDecoder[IRI] = Implicits.iriKeyDecoder(biolinkData.prefixes)
+
+        implicit val biolinkClassEncoder: Encoder[BiolinkClass] = Implicits.biolinkClassEncoder
+        implicit val biolinkClassDecoder: Decoder[BiolinkClass] = Implicits.biolinkClassDecoder(biolinkData.classes)
+
+        implicit val biolinkPredicateEncoder: Encoder[BiolinkPredicate] = Implicits.biolinkPredicateEncoder(biolinkData.prefixes)
+        implicit val biolinkPredicateDecoder: Decoder[List[BiolinkPredicate]] =
```

Comment:
Just a comment—we do this in a lot of places. We should make a class which we can construct from biolink data, whose objects provide implicit fields that can just be imported, like:

```scala
val de = DecoderEncoder(biolinkData)
import de._
```

Reply:
Merged master in 408d492e9a54bfc2 and replaced this in 83bf5e4.